### PR TITLE
Refactor/aggs metadata

### PR DESCRIPTION
**Pull request by gmarz**
Support aggregation meta data https://www.elastic.co/guide/en/elasticsearch/reference/2.1/agg-metadata.html (added by our very own @Mpdreamz  :) )

This PR also introduces another refactoring to aggregations which I think greatly simplifies the object graph (see https://github.com/elastic/elasticsearch-net/commit/8bdcbb3c27051175ec4e2f9865ea9e099b8c2014 and c509515).

### CODE REVIEW

File: src/Nest/Aggregations/Aggregation.cs [hunk 1]
```diff
@@ -1,32 +1,26 @@
...
-	/// </summary>
-	[ExactContractJsonConverter(typeof(AggregationJsonConverter))]
-	public interface IAggregation
-	{
-	}
-
-	/// <summary>
-	/// Base for the OIS Aggregation DSL
-	/// </summary>
-	public interface IAggregationBase : IAggregation
```
Comment (Mpdreamz) [inline src/Nest/Aggregations/Aggregation.cs]: I like this but I'm wondering why we didn't do this in the first place, do you know @gmarz?

File: src/Nest/Aggregations/AggregationResult.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Nest
+{
+	[ExactContractJsonConverter(typeof(AggregationResultJsonConverter))]
```
Comment (Mpdreamz) [inline src/Nest/Aggregations/AggregationResult.cs]: The JsonConverter name does not match the type its handling

File: src/Tests/tests.yaml [hunk 3]
```diff
@@ -1,5 +1,9 @@
 ﻿# mode either u (unit test), i (integration test) or m (mixed mode)
+<<<<<<< .mine
+mode: i
```
Comment (Mpdreamz) [inline src/Tests/tests.yaml]: Unresolved merge conflict here

File: src/Nest/Aggregations/Bucket/Bucket.cs [hunk 4]
```diff
@@ -1,35 +1,50 @@
...
-		public Bucket(IDictionary<string, IAggregation> aggregations) : base(aggregations) { }
+		public DocCountBucket() { }
+		public DocCountBucket(IDictionary<string, IAggregationItem> aggregations) : base(aggregations) { }
 
 		public IList<TBucketItem> Items { get; set; }
 	}
 
-	public abstract class BucketBase : AggregationsHelper, IBucket
+	// Intermediate object used for deserialization
+	internal class BucketDto : IAggregationResult
```
Comment (Mpdreamz) [inline src/Nest/Aggregations/Bucket/Bucket.cs]: I'm really hating DTO as a suffix but can not suggest a more meaningful name for this :smile:

File: src/Nest/Aggregations/Aggregation.cs (hunk 1 above)
Comment (gmarz) [inline src/Nest/Aggregations/Aggregation.cs]: I'm not really sure of original reasoning, but we didn't have to.  I think the new separation is much cleaner.

File: src/Nest/Aggregations/AggregationResult.cs (hunk 2 above)
Comment (gmarz) [inline src/Nest/Aggregations/AggregationResult.cs]: Yea, I'm having trouble coming up with a batter name for `IAggregationItem` and `IAggregationResult`.  `IAggregation` would be good, but that already exists for requests.

File: src/Tests/tests.yaml (hunk 3 above)
Comment (gmarz) [inline src/Tests/tests.yaml]: doh, thanks

File: src/Nest/Aggregations/Bucket/Bucket.cs (hunk 4 above)
Comment (gmarz) [inline src/Nest/Aggregations/Bucket/Bucket.cs]: Likewise.  The entire thing is a code smell IMO...I've made it internal so now we can work on getting rid of it entirely by refactoring the json converter bits.

**Comment by russcam** [comment]
Reviewing now...

**Comment by russcam** [comment]
LGTM :+1:

**Comment by RobinRieger** [comment]
I just downloaded the latest code and it gives me some compile errors re the `DocCount` after I changed all the references from `items` to `buckets`. I think `DocCount` used to be a `long` in the `SingleBucket`/`KeyedBucket` but is now a `long?` in the new `KeyedBucketItem`. 

Is it intentional to be nullable? When will `DocCount` be null? If it can't find anything shouldn't it be zero?

**Comment by gmarz** [comment]
@RobinRieger making them nullable made testing easier, but you're right, they shouldn't be.  Will fix this for the next release, thanks :+1:

**Pull request merged**